File: Sample/EventStoreDB/Simple/ECommerce/Pricing/ProductPricing/RandomProductPriceCalculator.cs
using System;
using System.Collections.Concurrent;
using ECommerce.ShoppingCarts.ProductItems;

namespace ECommerce.Pricing.ProductPricing;

public class RandomProductPriceCalculator: IProductPriceCalculator
{
    private readonly ConcurrentDictionary<Guid, decimal> productPrices = new();

    public PricedProductItem Calculate(ProductItem productItem)
    {
        var random = new Random();

        var price = productPrices.GetOrAdd(
            productItem.ProductId,
            (decimal)random.NextDouble() * 100
        );
        return PricedProductItem.From(productItem, price);
    }
}

File: Sample/ECommerce.Equinox/ECommerce.Domain/RandomProductPriceCalculator.fs
namespace ECommerce.Domain

type RandomProductPriceCalculator() =

    let productPrices = System.Collections.Concurrent.ConcurrentDictionary<ProductId, decimal>()

    member _.Calculate(productId : ProductId) : Async<decimal> = async {
        let r = System.Random()
        let calc _ = (r.NextDouble() |> decimal) * 100m
        let price : decimal = productPrices.GetOrAdd(productId, calc)
        return price
    }
